# Transformación de DataFrames

## Inspeccionar un DataFrame

In [7]:
# Importar la libreria de Pandas
import pandas as pd

In [8]:
homelessness = pd.read_csv('./Data_Sets/homelessness.csv')

In [9]:
# Imprimir los primeros datos con head
homelessness.head()

Unnamed: 0              region       state  individuals  family_members  \
0           0  East South Central     Alabama       2570.0           864.0   
1           1             Pacific      Alaska       1434.0           582.0   
2           2            Mountain     Arizona       7259.0          2606.0   
3           3  West South Central    Arkansas       2280.0           432.0   
4           4             Pacific  California     109008.0         20964.0   

   state_pop  
0    4887681  
1     735139  
2    7158024  
3    3009733  
4   39461588

In [10]:
# Imprimir informacion acerca de homelessness
print(homelessness.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      51 non-null     int64  
 1   region          51 non-null     object 
 2   state           51 non-null     object 
 3   individuals     51 non-null     float64
 4   family_members  51 non-null     float64
 5   state_pop       51 non-null     int64  
dtypes: float64(2), int64(2), object(2)
memory usage: 2.5+ KB
None


In [11]:
print(homelessness.shape)

(51, 6)


In [12]:
# Imprimir la descripcion de homelessness
print(homelessness.describe())

       Unnamed: 0    individuals  family_members     state_pop
count   51.000000      51.000000       51.000000  5.100000e+01
mean    25.000000    7225.784314     3504.882353  6.405637e+06
std     14.866069   15991.025083     7805.411811  7.327258e+06
min      0.000000     434.000000       75.000000  5.776010e+05
25%     12.500000    1446.500000      592.000000  1.777414e+06
50%     25.000000    3082.000000     1482.000000  4.461153e+06
75%     37.500000    6781.500000     3196.000000  7.340946e+06
max     50.000000  109008.000000    52070.000000  3.946159e+07


## Partes de un DataFrame

In [13]:
print(homelessness.values)

[[0 'East South Central' 'Alabama' 2570.0 864.0 4887681]
 [1 'Pacific' 'Alaska' 1434.0 582.0 735139]
 [2 'Mountain' 'Arizona' 7259.0 2606.0 7158024]
 [3 'West South Central' 'Arkansas' 2280.0 432.0 3009733]
 [4 'Pacific' 'California' 109008.0 20964.0 39461588]
 [5 'Mountain' 'Colorado' 7607.0 3250.0 5691287]
 [6 'New England' 'Connecticut' 2280.0 1696.0 3571520]
 [7 'South Atlantic' 'Delaware' 708.0 374.0 965479]
 [8 'South Atlantic' 'District of Columbia' 3770.0 3134.0 701547]
 [9 'South Atlantic' 'Florida' 21443.0 9587.0 21244317]
 [10 'South Atlantic' 'Georgia' 6943.0 2556.0 10511131]
 [11 'Pacific' 'Hawaii' 4131.0 2399.0 1420593]
 [12 'Mountain' 'Idaho' 1297.0 715.0 1750536]
 [13 'East North Central' 'Illinois' 6752.0 3891.0 12723071]
 [14 'East North Central' 'Indiana' 3776.0 1482.0 6695497]
 [15 'West North Central' 'Iowa' 1711.0 1038.0 3148618]
 [16 'West North Central' 'Kansas' 1443.0 773.0 2911359]
 [17 'East South Central' 'Kentucky' 2735.0 953.0 4461153]
 [18 'West South Cen

In [14]:
# Imprimir los indices de las columnas  de homelessness
print(homelessness.columns)

Index(['Unnamed: 0', 'region', 'state', 'individuals', 'family_members',
       'state_pop'],
      dtype='object')


In [15]:
# Imprimir los indices de las filas
print(homelessness.index)

RangeIndex(start=0, stop=51, step=1)


## Ordenar Filas
| Ordenar en ...| Sintaxis                  |
|---------------| --------------------------|
| Una Columna   | `df.sort_values("bread")` |
| Varias Columnas   | `df.sort_values(["bread", "weight_kg"])` |

1. Ordena homelessness por el numero de personas sin hogar de la columna individuals, de menor a mayor, y guardarlo como homelessness_ind

1. Ordena homelessness por el numero de personas sin hogar  family_members en orden descendente y guardarlo como homelessness_fam

1. Ordenar homelessness primero por region (ascendente) y luego por numero de miembros de la familia (descendente). Guardarlo como homelessness_reg_fam

In [16]:
homelessness_ind = homelessness.sort_values('individuals')
print(homelessness_ind.head(10))

    Unnamed: 0              region          state  individuals  \
50          50            Mountain        Wyoming        434.0   
34          34  West North Central   North Dakota        467.0   
7            7      South Atlantic       Delaware        708.0   
39          39         New England   Rhode Island        747.0   
45          45         New England        Vermont        780.0   
29          29         New England  New Hampshire        835.0   
41          41  West North Central   South Dakota        836.0   
26          26            Mountain        Montana        983.0   
48          48      South Atlantic  West Virginia       1021.0   
24          24  East South Central    Mississippi       1024.0   

    family_members  state_pop  
50           205.0     577601  
34            75.0     758080  
7            374.0     965479  
39           354.0    1058287  
45           511.0     624358  
29           615.0    1353465  
41           323.0     878698  
26           422.

In [17]:
# Ordenar homelessness por family_members de forma descendente
homelessness_fam = homelessness.sort_values('family_members', ascending=False)
print(homelessness_fam.head(10))

    Unnamed: 0              region          state  individuals  \
32          32        Mid-Atlantic       New York      39827.0   
4            4             Pacific     California     109008.0   
21          21         New England  Massachusetts       6811.0   
9            9      South Atlantic        Florida      21443.0   
43          43  West South Central          Texas      19199.0   
47          47             Pacific     Washington      16424.0   
38          38        Mid-Atlantic   Pennsylvania       8163.0   
13          13  East North Central       Illinois       6752.0   
30          30        Mid-Atlantic     New Jersey       6048.0   
37          37             Pacific         Oregon      11139.0   

    family_members  state_pop  
32         52070.0   19530351  
4          20964.0   39461588  
21         13257.0    6882635  
9           9587.0   21244317  
43          6111.0   28628666  
47          5880.0    7523869  
38          5349.0   12800922  
13          3891.

In [18]:
homelessness_reg_fam = homelessness.sort_values(['region', 'family_members'], ascending= [True, False])
print(homelessness_reg_fam.head(20))

    Unnamed: 0              region         state  individuals  family_members  \
13          13  East North Central      Illinois       6752.0          3891.0   
35          35  East North Central          Ohio       6929.0          3320.0   
22          22  East North Central      Michigan       5209.0          3142.0   
49          49  East North Central     Wisconsin       2740.0          2167.0   
14          14  East North Central       Indiana       3776.0          1482.0   
42          42  East South Central     Tennessee       6139.0          1744.0   
17          17  East South Central      Kentucky       2735.0           953.0   
0            0  East South Central       Alabama       2570.0           864.0   
24          24  East South Central   Mississippi       1024.0           328.0   
32          32        Mid-Atlantic      New York      39827.0         52070.0   
38          38        Mid-Atlantic  Pennsylvania       8163.0          5349.0   
30          30        Mid-At

## Subconjuntos de Columnas 

Hay muchas formas de subconjuntar un DataFrame, quiza la mas comun sea utilizar  operadores relacionales devolver True o False para cada fila y luego pasarlo entre corchetes

`dogs[dogs['height_cm']>60]`
`dogs[dogs['color'] == 'tan']`

Puedes filtrar por varias condiciones a la vez utilizando el operador & 

`dogs[(dogs['heigth_cm'] > 60) & (dogs['color'] == 'tan')]`

**Instrucciones**

1. Filtra homelessness para los casos en los que el numero de individuals sea superior a diez mil, asignandolos a ind_gt_10k. Visualizar el resultado

2. Filtra homelessness para los casos que la region es Mountain, asignado a mountain_reg. Visualiza el resultado

3. Filtra homelessness para los casos en los que el numero de family_members sea inferior a mil y la region sea 'Pacifico', asignandolo a fam_lt_1k_pac. Visualizar el resultado

In [19]:
ind_gt_10k = homelessness[homelessness['individuals'] > 10000]
print(ind_gt_10k.head(10))

    Unnamed: 0              region       state  individuals  family_members  \
4            4             Pacific  California     109008.0         20964.0   
9            9      South Atlantic     Florida      21443.0          9587.0   
32          32        Mid-Atlantic    New York      39827.0         52070.0   
37          37             Pacific      Oregon      11139.0          3337.0   
43          43  West South Central       Texas      19199.0          6111.0   
47          47             Pacific  Washington      16424.0          5880.0   

    state_pop  
4    39461588  
9    21244317  
32   19530351  
37    4181886  
43   28628666  
47    7523869  


In [20]:
mountain_reg = homelessness[homelessness['region'] == 'Mountain']
mountain_reg.sort_values('region', ascending= False)
print(mountain_reg.head(10))

    Unnamed: 0    region       state  individuals  family_members  state_pop
2            2  Mountain     Arizona       7259.0          2606.0    7158024
5            5  Mountain    Colorado       7607.0          3250.0    5691287
12          12  Mountain       Idaho       1297.0           715.0    1750536
26          26  Mountain     Montana        983.0           422.0    1060665
28          28  Mountain      Nevada       7058.0           486.0    3027341
31          31  Mountain  New Mexico       1949.0           602.0    2092741
44          44  Mountain        Utah       1904.0           972.0    3153550
50          50  Mountain     Wyoming        434.0           205.0     577601


In [21]:
fam_lt_1k_pac = homelessness[(homelessness['family_members'] < 1000) & (homelessness['region'] == 'Pacific')]
print(fam_lt_1k_pac.head(10))

   Unnamed: 0   region   state  individuals  family_members  state_pop
1           1  Pacific  Alaska       1434.0           582.0     735139


## Subconjunto de Filas por variables categoricas

Utiliza el metodo `.isin()` que permitira abordar el problema escribiendo en una sola condicion en lugar de varias por que  para segmentar variables categoricas se utiliza el operador pipe `|` 

`Colors = ['brown', 'black', 'tan']`

`condition = dogs['color'].isin(colors)`

`dogs['condition']`

**Instrucciones**

1. Filtra homelessness para los casos en los que el censo para los estados de California, Arizona, Nevada y Utha, asignandolo a mojave_homelessness

In [22]:
canu = ['California', 'Arizona', 'Nevada', 'Utah']
mojave_homelessness = homelessness[homelessness['state'].isin(canu)]

print(mojave_homelessness)

    Unnamed: 0    region       state  individuals  family_members  state_pop
2            2  Mountain     Arizona       7259.0          2606.0    7158024
4            4   Pacific  California     109008.0         20964.0   39461588
28          28  Mountain      Nevada       7058.0           486.0    3027341
44          44  Mountain        Utah       1904.0           972.0    3153550


## Añadir dos nuevas columnas

No te quedes solo con los datos que te dan. En su lugar, se pueden añadir nuevas columnas aun DataFrame. SA esto se le llama `Transformación`, `Mutación` e `Ingeniería de Caracteristicas`.

Se pueden crear columnas nuevas desde cero, pues es habitula derivarlas de otras columnas, por ejemplo, sumando columnas o cambiando sus unidades.

**Instrucciones**

- Añade una nueva columna a homelessness, denominada `total`, que contenga la suma de las columnas individuales y family_members
- Añade otra columna a homelessness, denominada p_homeless, que contenga la proporción de la población de personas sin hogar total con respecto a la población total de cada estado stage_pop.

In [26]:
# Agregar una columna total como la suma de individuals y family_members
homelessness['total'] = homelessness['individuals'] + homelessness['family_members']

# Agregar una columna p_homeless como proporción del total de homeless
# de la población para el estado de la población (state_pop)

homelessness['p_homeless'] = homelessness['total'] / homelessness['state_pop']

# Visualizar el resultado
print(homelessness.head(10))

   Unnamed: 0              region                 state  individuals  \
0           0  East South Central               Alabama       2570.0   
1           1             Pacific                Alaska       1434.0   
2           2            Mountain               Arizona       7259.0   
3           3  West South Central              Arkansas       2280.0   
4           4             Pacific            California     109008.0   
5           5            Mountain              Colorado       7607.0   
6           6         New England           Connecticut       2280.0   
7           7      South Atlantic              Delaware        708.0   
8           8      South Atlantic  District of Columbia       3770.0   
9           9      South Atlantic               Florida      21443.0   

   family_members  state_pop     total  p_homeless  
0           864.0    4887681    3434.0    0.000703  
1           582.0     735139    2016.0    0.002742  
2          2606.0    7158024    9865.0    0.0013

In [30]:
homelessness[['state', 'individuals', 'state_pop', 'total', 'p_homeless']].head(10)

state  individuals  state_pop     total  p_homeless
0               Alabama       2570.0    4887681    3434.0    0.000703
1                Alaska       1434.0     735139    2016.0    0.002742
2               Arizona       7259.0    7158024    9865.0    0.001378
3              Arkansas       2280.0    3009733    2712.0    0.000901
4            California     109008.0   39461588  129972.0    0.003294
5              Colorado       7607.0    5691287   10857.0    0.001908
6           Connecticut       2280.0    3571520    3976.0    0.001113
7              Delaware        708.0     965479    1082.0    0.001121
8  District of Columbia       3770.0     701547    6904.0    0.009841
9               Florida      21443.0   21244317   31030.0    0.001461

In [29]:
columnas = ['state', 'individuals', 'state_pop', 'total', 'p_homeless']
homelessness[columnas]

state  individuals  state_pop     total  p_homeless
0                Alabama       2570.0    4887681    3434.0    0.000703
1                 Alaska       1434.0     735139    2016.0    0.002742
2                Arizona       7259.0    7158024    9865.0    0.001378
3               Arkansas       2280.0    3009733    2712.0    0.000901
4             California     109008.0   39461588  129972.0    0.003294
5               Colorado       7607.0    5691287   10857.0    0.001908
6            Connecticut       2280.0    3571520    3976.0    0.001113
7               Delaware        708.0     965479    1082.0    0.001121
8   District of Columbia       3770.0     701547    6904.0    0.009841
9                Florida      21443.0   21244317   31030.0    0.001461
10               Georgia       6943.0   10511131    9499.0    0.000904
11                Hawaii       4131.0    1420593    6530.0    0.004597
12                 Idaho       1297.0    1750536    2012.0    0.001149
13              Illinois       6752.0   12723071   10643.0    0.000837
14               Indiana       3776.0    6695497    5258.0    0.000785
15                  Iowa       1711.0    3148618    2749.0    0.000873
16                Kansas       1443.0    2911359    2216.0    0.000761
17              Kentucky       2735.0    4461153    3688.0    0.000827
18             Louisiana       2540.0    4659690    3059.0    0.000656
19                 Maine       1450.0    1339057    2516.0    0.001879
20              Maryland       4914.0    6035802    7144.0    0.001184
21         Massachusetts       6811.0    6882635   20068.0    0.002916
22              Michigan       5209.0    9984072    8351.0    0.000836
23             Minnesota       3993.0    5606249    7243.0    0.001292
24           Mississippi       1024.0    2981020    1352.0    0.000454
25              Missouri       3776.0    6121623    5883.0    0.000961
26               Montana        983.0    1060665    1405.0    0.001325
27              Nebraska       1745.0    1925614    2421.0    0.001257
28                Nevada       7058.0    3027341    7544.0    0.002492
29         New Hampshire        835.0    1353465    1450.0    0.001071
30            New Jersey       6048.0    8886025    9398.0    0.001058
31            New Mexico       1949.0    2092741    2551.0    0.001219
32              New York      39827.0   19530351   91897.0    0.004705
33        North Carolina       6451.0   10381615    9268.0    0.000893
34          North Dakota        467.0     758080     542.0    0.000715
35                  Ohio       6929.0   11676341   10249.0    0.000878
36              Oklahoma       2823.0    3940235    3871.0    0.000982
37                Oregon      11139.0    4181886   14476.0    0.003462
38          Pennsylvania       8163.0   12800922   13512.0    0.001056
39          Rhode Island        747.0    1058287    1101.0    0.001040
40        South Carolina       3082.0    5084156    3933.0    0.000774
41          South Dakota        836.0     878698    1159.0    0.001319
42             Tennessee       6139.0    6771631    7883.0    0.001164
43                 Texas      19199.0   28628666   25310.0    0.000884
44                  Utah       1904.0    3153550    2876.0    0.000912
45               Vermont        780.0     624358    1291.0    0.002068
46              Virginia       3928.0    8501286    5975.0    0.000703
47            Washington      16424.0    7523869   22304.0    0.002964
48         West Virginia       1021.0    1804291    1243.0    0.000689
49             Wisconsin       2740.0    5807406    4907.0    0.000845
50               Wyoming        434.0     577601     639.0    0.001106

## Practica - Combo Ataque

Hasta ahorita se han visto los 4 tipos más comunes de manipulación de datos: `ordenar filas`, `subconjuntar columnas`, `subconjuntar filas` y `Añadir nuevas columnas`.

En un analisis de la vida real puedes meztlar y conbinar estas 4 manipulaciones para responder a multiples preguntas.

En este ejercisio, responderas la pregunta: `"¿Qué estado tiene el mayor numero de personas in hogar por cada 10,000 habitantes"`. Convina las nuevas habilidades en pandas para averiguarlo

**Intrucciones**

- Añade una columna a homelessness, indiv_per_10k, que contenga el numero de personas sin hogar por cada 10,000 de personas en cada estado, utilizando state_pop para la población del estado
- Subconjunta las filas en las que el indiv_per_10k sea superiora 20 asignandolas a high_homelessness
- Ordena high_homelessness en orden desendente indiv_per_10k, asigando a high_homelessness_srt. Seleccionando solo las columnas state e indiv_per_10k de high_homelessness_srt y guardar como result y Visualiza result

In [31]:
# Crear la columna indiv_per_10k

homelessness['indiv_per_10k'] =  10000 * homelessness ['individuals'] / homelessness['state_pop']

# Subconjuntar filas para indiv_per_10k mayores que 20

high_homelessness = homelessness[homelessness['indiv_per_10k']>20]

# Ordenar hight_homelessness por u¿indiv_per_10k
result = high_homelessness.sort_values('indiv_per_10k', ascending=False)

# Visualizar el resultado
result

Unnamed: 0          region                 state  individuals  \
8            8  South Atlantic  District of Columbia       3770.0   
11          11         Pacific                Hawaii       4131.0   
4            4         Pacific            California     109008.0   
37          37         Pacific                Oregon      11139.0   
28          28        Mountain                Nevada       7058.0   
47          47         Pacific            Washington      16424.0   
32          32    Mid-Atlantic              New York      39827.0   

    family_members  state_pop     total  p_homeless  indiv_per_10k  
8           3134.0     701547    6904.0    0.009841      53.738381  
11          2399.0    1420593    6530.0    0.004597      29.079406  
4          20964.0   39461588  129972.0    0.003294      27.623825  
37          3337.0    4181886   14476.0    0.003462      26.636307  
28           486.0    3027341    7544.0    0.002492      23.314189  
47          5880.0    7523869   22304.0    0.002964      21.829195  
32         52070.0   19530351   91897.0    0.004705      20.392363

El District of columbia tiene el mayor número de individuos sin hogar - casi el 54% por cada 10,000 personas. esto es casi el doble del número del siguiente estado con mas casos, Hawaii